### ISSUE TITLE
LcfReader::ReadString heap corruption on maliciously crafted files.

### ISSUE THREAD

**Issue by scrippie** [EasyRPG/liblcf#194]
_LcfReader::ReadString()_ can corrupt the heap if the size calculation _size + 1_ wraps around. This allocates a region of size 0, while copying more data than this space can hold.

Below a simple test to trigger the problem through LDB_Reader which reads a signed integer of -1 for the size, which is converted to a _size_t,_ resulting in heap corruption on my Linux system.

It's probably a more safe idea to error in this situation, rather than opening the door for execution of arbitrary code through malicious LDB files.

```c++
#include <cstdio>
#include <cstdlib>
#include "ldb_reader.h"
#include "reader_lcf.h"
#include "writer_lcf.h"

void build_test_ldb(void)
{
        LcfWriter writer("test_ldb_reader.bin", "UTF-8");

        writer.WriteInt(-1);

        for (unsigned int i = 0; i < 10000; i++)
                writer.Write((uint8_t)'A');
}

int main(void)
{
        build_test_ldb();

        if (!LDB_Reader::Load("test_ldb_reader.bin", "UTF-8")) {
                fprintf(stderr, "%s\n", LcfReader::GetError().c_str());
                exit(EXIT_FAILURE);
        }

        exit(EXIT_SUCCESS);
}

```

Proposed **untested** patch is below. This does not add space for the terminating 0-byte, as it is unnecessary for the _std::string_ constructor called. This also allows 0-bytes to be part of the string, which is arguably more sensible for length encoded rather than 0-terminated strings.

```patch
diff --git a/src/reader_lcf.cpp b/src/reader_lcf.cpp
index b3141bb..3e07e99 100644
--- a/src/reader_lcf.cpp
+++ b/src/reader_lcf.cpp
@@ -160,10 +160,9 @@ void LcfReader::Read<uint32_t>(std::vector<uint32_t> &buffer, size_t size) {
 }
 
 void LcfReader::ReadString(std::string& ref, size_t size) {
-       char* chars = new char[size + 1];
-       chars[size] = '\0';
+       char* chars = new char[size];
        Read(chars, 1, size);
-       ref = Encode(std::string(chars));
+       ref = Encode(std::string(chars, size));
        delete[] chars;
 }

```

**Comment by Ghabry** [comment]
Thanks for the patch. Please note that we currently don't care much about security in liblcf and Player. I assume that it is quite easy to create a malicious ldb file which generates out of bounds reads and writes in the Player due to missing range checks.

But bugs that allow the exploitation of liblcf directly are of course worse because the lib is used by multiple programs. If you find any other vulnerable spots please report them (but not normal crashes, there are probably hundreds ;) )

**Comment by scrippie** [comment]
Yeah, I figured security wasn't high on the priority list at the moment. I was giving the project a read out of interest in the engine itself, but come from a security background so I always keep my eyes open for such issues.

I'll submit similar issues when I run into them, but for now will only bother with the ones that seem easy to fix and affect a small localized region in the code. That should be relatively low effort for you guys to deal with, while it hardens the engine a little for the future.

**Comment by Ghabry** [comment]
Looks like a good candidate to fix before 0.5.1 ^^
